pyroomacoustics/directivities.py

```diff
@@ -0,0 +1,183 @@
+import numpy as np
+from pyroomacoustics.utilities import Options
+
+
+class DirectivityPattern(Options):
+    FIGURE_EIGHT = "figure_eight"
+    HYPERCARDIOID = "hypercardioid"
+    CARDIOID = "cardioid"
+    SUBCARDIOID = "subcardioid"
+    OMNI = "omni"
+
+
+PATTERN_TO_CONSTANT = {
+    DirectivityPattern.FIGURE_EIGHT: 0,
+    DirectivityPattern.HYPERCARDIOID: 0.25,
+    DirectivityPattern.CARDIOID: 0.5,
+    DirectivityPattern.SUBCARDIOID: 0.75,
+    DirectivityPattern.OMNI: 1.0
+}
+
+
+class DirectionVector(object):
+    """
+    Parameters
+    ----------
+    azimuth : float
+    colatitude : float, optional
+        Default to PI / 2, only XY plane.
+    degrees : bool
+        Whether provided values are in degrees (True) or radians (False).
+    """
+    def __init__(self, azimuth, colatitude=None, degrees=True):
+        if degrees is True:
+            azimuth = np.radians(azimuth)
+            if colatitude is not None:
+                colatitude = np.radians(colatitude)
+        self._azimuth = azimuth
+        if colatitude is None:
+            colatitude = np.pi / 2
+        assert colatitude <= np.pi and colatitude >= 0
+        self._colatitude = colatitude
+
+    def get_azimuth(self, degrees=False):
+        if degrees:
+            return np.degrees(self._azimuth)
+        else:
+            return self._azimuth
+
+    def get_colatitude(self, degrees=False):
+        if degrees:
+            return np.degrees(self._colatitude)
+        else:
+            return self._colatitude
+
+
+class Directivity(object):
```

Comment:
I think `Directivity` is a little too general for this class since it refers only to the Cardioid/Hypercardioid, etc family of patterns. Does this family have a name ?

Also, I think it would be nice to create derived classes that we directly call for the different types.
```python
class Cardioid(Directivity):
    def __init__(self, orientation=None, ...)
        super().__init__(orientation=orientation, pattern=DirectivityPattern.CARDIOID)
```

Reply:
Maybe the class can be called `CardioidFamily`.

Reply:
And, can we make it also a simple function, that may come in handy for what I am currently doing with the ray tracing.
```python
def cardioid_func(x, loc, coef, gain):
     """
     Parameters
     x: array_like, shape (..., n_dim)
          The input
     loc: array_like, shape (n_dim)
          The direction vector (should be normalized, but no need to check in the function)
     coeff: float
          The parameter of the cardioid
     gain: float
          The gain
     """
     return gain * (coeff + (1 - coeff) * np.matmul(x, loc)
```
And we will also need to have functions for the volume (3D) and area (2D) of cardioids for the energy matching.

Reply:
done in most recent commits!

Reply:
Cool! I am now using this! Small detail, the shape of `x` in the docstring doesn't match the implementation. The docstring should say `(n_dim, n_points)` or something like that.